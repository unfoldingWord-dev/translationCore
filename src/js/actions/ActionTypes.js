/* istanbul lignore next */
const ActionTypes = {
  CHANGE_ONLINE_STATUS: "CHANGE_ONLINE_STATUS",
  START_LOADING: "START_LOADING",
  SHOW_MODAL_CONTAINER: "SHOW_MODAL_CONTAINER",
  SELECT_MODAL_TAB: "SELECT_MODAL_TAB",
  LOGOUT_USER: "LOGOUT_USER",
  LOGIN_USER: "LOGIN_USER",
  SET_TOOLS_METADATA: "SET_TOOLS_METADATA",
  GET_RECENT_PROJECTS: "GET_RECENT_PROJECTS",
  RECIEVE_REPOS: "RECIEVE_REPOS",
  IMPORT_LINK: "IMPORT_LINK",
  TOGGLE_LOADER_MODAL: "TOGGLE_LOADER_MODAL",
  SELECT_MODAL_SECTION: "SELECT_MODAL_SECTION",
  SHOW_POPOVER: "SHOW_POPOVER",
  CLOSE_POPOVER: "CLOSE_POPOVER",
  SAVE_TOOL_VIEW: "SAVE_TOOL_VIEW",
  // Resources Actions
  ADD_NEW_BIBLE_TO_RESOURCES: "ADD_NEW_BIBLE_TO_RESOURCES",
  ADD_TRANSLATIONHELPS_ARTICLE: "ADD_TRANSLATIONHELPS_ARTICLE",
  ADD_LEXICON_ENTRY: "ADD_LEXICON_ENTRY",
  // Other actions that need organized
  SET_SAVE_PATH_LOCATION: "SET_SAVE_PATH_LOCATION",
  UPDATE_PROGRESS: "UPDATE_PROGRESS",
  TOGGLE_REMINDER: "TOGGLE_REMINDER",
  SET_REMINDER: "SET_REMINDER",
  CHANGE_CURRENT_CONTEXT_ID: "CHANGE_CURRENT_CONTEXT_ID",
  ADD_COMMENT: "ADD_COMMENT",
  ADD_VERSE_EDIT: "ADD_VERSE_EDIT",
  ADD_GROUP_DATA: "ADD_GROUP_DATA",
  CHANGE_SELECTIONS: "CHANGE_SELECTIONS",
  CLEAR_CURRENT_TOOL_DATA: "CLEAR_CURRENT_TOOL_DATA",
  STORE_MANIFEST: "STORE_MANIFEST",
  STORE_SAVE_LOCATION: "STORE_SAVE_LOCATION",
  SET_CURRENT_TOOL_NAME: "SET_CURRENT_TOOL_NAME",
  SET_CURRENT_TOOL_TITLE: "SET_CURRENT_TOOL_TITLE",
  LOAD_GROUPS_INDEX: "LOAD_GROUPS_INDEX",
  LOAD_GROUPS_DATA_FROM_FS: "LOAD_GROUPS_DATA_FROM_FS",
  SET_REMINDERS_IN_GROUPDATA: "SET_REMINDERS_IN_GROUPDATA",
  SET_INVALIDATION_IN_GROUPDATA: "SET_INVALIDATION_IN_GROUPDATA",
  TOGGLE_REMINDERS_IN_GROUPDATA: "TOGGLE_REMINDERS_IN_GROUPDATA",
  TOGGLE_SELECTIONS_IN_GROUPDATA: "TOGGLE_SELECTIONS_IN_GROUPDATA",
  TOGGLE_VERSE_EDITS_IN_GROUPDATA: "TOGGLE_VERSE_EDITS_IN_GROUPDATA",
  TOGGLE_COMMENTS_IN_GROUPDATA: "TOGGLE_COMMENTS_IN_GROUPDATA",
  CLEAR_PREVIOUS_GROUPS_DATA: "CLEAR_PREVIOUS_GROUPS_DATA",
  CLEAR_PREVIOUS_GROUPS_INDEX: "CLEAR_PREVIOUS_GROUPS_INDEX",
  CLEAR_CONTEXT_ID: "CLEAR_CONTEXT_ID",
  UPDATE_TARGET_VERSE: "UPDATE_TARGET_VERSE",
  SHOW_LOADING_CIRCLE: "SHOW_LOADING_CIRCLE",
  HIDE_LOADING_CIRCLE: "HIDE_LOADING_CIRCLE",
  GOGS_SERVER_ERROR: "GOGS_SERVER_ERROR",
  OPEN_ALERT_DIALOG: "OPEN_ALERT_DIALOG",
  OPEN_OPTION_DIALOG: "OPEN_OPTION_DIALOG",
  CLOSE_ALERT_DIALOG: "CLOSE_ALERT_DIALOG",
  LOGIN_LOCAL_USER: "LOGIN_LOCAL_USER",
  RESET_PROJECT_DETAIL: "RESET_PROJECT_DETAIL",
  CLEAR_RESOURCES_REDUCER: "CLEAR_RESOURCES_REDUCER",
  UPDATE_ONLINE_MODE: "UPDATE_ONLINE_MODE",
  // Online import actions
  SET_REPOS_DATA: "SET_REPOS_DATA",
  // project actions
  SET_PROJECT_PROGRESS_FOR_TOOL: "SET_PROJECT_PROGRESS_FOR_TOOL",
  SET_GL_FOR_TOOL: "SET_GL_FOR_TOOL",
  // Home
  TOGGLE_HOME_VIEW: "TOGGLE_HOME_VIEW",
  TOGGLE_WELCOME_SPLASH: "TOGGLE_WELCOME_SPLASH",
  GO_TO_STEP: "GO_TO_STEP",
  OPEN_PROJECTS_FAB: "OPEN_PROJECTS_FAB",
  CLOSE_PROJECTS_FAB: "CLOSE_PROJECTS_FAB",
  GET_MY_PROJECTS: "GET_MY_PROJECTS",
  OPEN_ONLINE_IMPORT_MODAL: "OPEN_ONLINE_IMPORT_MODAL",
  CLOSE_ONLINE_IMPORT_MODAL: "CLOSE_ONLINE_IMPORT_MODAL",
  OPEN_LICENSE_MODAL: "OPEN_LICENSE_MODAL",
  CLOSE_LICENSE_MODAL: "CLOSE_LICENSE_MODAL",
  RESET_IMPORT_ONLINE_REDUCER: "RESET_IMPORT_ONLINE_REDUCER",
  SHOW_DIMMED_SCREEN: "SHOW_DIMMED_SCREEN",
  //Project Validation
  RESET_PROJECT_VALIDATION_REDUCER: "RESET_PROJECT_VALIDATION_REDUCER",
  ONLY_SHOW_PROJECT_INFORMATION_SCREEN: "ONLY_SHOW_PROJECT_INFORMATION_SCREEN",
  TOGGLE_PROJECT_VALIDATION_STEPPER: "TOGGLE_PROJECT_VALIDATION_STEPPER",
  GO_TO_PROJECT_VALIDATION_STEP: "GO_TO_PROJECT_VALIDATION_STEP",
  UPDATE_PROJECT_VALIDATION_NEXT_BUTTON_STATUS: "UPDATE_PROJECT_VALIDATION_NEXT_BUTTON_STATUS",
  MISSING_VERSES_CHECK: "MISSING_VERSES_CHECK",
  ADD_PROJECT_VALIDATION_STEP: "ADD_PROJECT_VALIDATION_STEP",
  REMOVE_PROJECT_VALIDATION_STEP: "REMOVE_PROJECT_VALIDATION_STEP",
  // project Details reducer
  UPDATE_SELECTED_PROJECT_FILENAME: "UPDATE_SELECTED_PROJECT_FILENAME",
  UPDATE_SOURCE_PROJECT_PATH: "UPDATE_SOURCE_PROJECT_PATH",
  SAVE_BOOK_ID_AND_BOOK_NAME_IN_MANIFEST: "SAVE_BOOK_ID_AND_BOOK_NAME_IN_MANIFEST",
  SAVE_RESOURCE_ID_IN_MANIFEST: "SAVE_RESOURCE_ID_IN_MANIFEST",
  SAVE_NICKNAME_IN_MANIFEST: "SAVE_NICKNAME_IN_MANIFEST",
  SAVE_LANGUAGE_DETAILS_IN_MANIFEST: "SAVE_LANGUAGE_DETAILS_IN_MANIFEST",
  SAVE_TRANSLATORS_LIST_IN_MANIFEST: "SAVE_TRANSLATORS_LIST_IN_MANIFEST",
  SAVE_CHECKERS_LIST_IN_MANIFEST: "SAVE_CHECKERS_LIST_IN_MANIFEST",
  ADD_MANIFEST_PROPERTY: "ADD_MANIFEST_PROPERTY",
  SET_PROJECT_TYPE: "SET_PROJECT_TYPE",
  OLD_SELECTED_PROJECT_FILENAME: "OLD_SELECTED_PROJECT_FILENAME",
  // Group menu
  TOGGLE_MENU_DRAWER: "TOGGLE_MENU_DRAWER",
  GROUP_MENU_EXPAND_SUBMENU: "GROUP_MENU_EXPAND_SUBMENU",
  GROUP_MENU_SET_FILTER: "GROUP_MENU_TOGGLE_FILTER",
  CLEAR_PREVIOUS_FILTERS: "CLEAR_PREVIOUS_FILTERS",
  // copy right check reducer
  SELECT_PROJECT_LICENSE_ID: "SELECT_PROJECT_LICENSE_ID",
  LOAD_PROJECT_LICENSE_MARKDOWN: "LOAD_PROJECT_LICENSE_MARKDOWN",
  CLEAR_COPYRIGHT_CHECK_REDUCER: "CLEAR_COPYRIGHT_CHECK_REDUCER",
  // project information check reducer
  SET_BOOK_ID_IN_PROJECT_INFORMATION_REDUCER: "SET_BOOK_ID_IN_PROJECT_INFORMATION_REDUCER",
  SET_RESOURCE_ID_IN_PROJECT_INFORMATION_REDUCER: "SET_RESOURCE_ID_IN_PROJECT_INFORMATION_REDUCER",
  SET_NICKNAME_IN_PROJECT_INFORMATION_REDUCER: "SET_NICKNAME_IN_PROJECT_INFORMATION_REDUCER",
  SET_LANGUAGE_ID_IN_PROJECT_INFORMATION_REDUCER: "SET_LANGUAGE_ID_IN_PROJECT_INFORMATION_REDUCER",
  SET_LANGUAGE_NAME_IN_PROJECT_INFORMATION_REDUCER: "SET_LANGUAGE_NAME_IN_PROJECT_INFORMATION_REDUCER",
  SET_LANGUAGE_DIRECTION_IN_PROJECT_INFORMATION_REDUCER: "SET_LANGUAGE_DIRECTION_IN_PROJECT_INFORMATION_REDUCER",
  SET_ALL_LANGUAGE_INFO_IN_PROJECT_INFORMATION_REDUCER: "SET_ALL_LANGUAGE_INFO_IN_PROJECT_INFORMATION_REDUCER",
  SET_CONTRIBUTORS_IN_PROJECT_INFORMATION_REDUCER: "SET_CONTRIBUTORS_IN_PROJECT_INFORMATION_REDUCER",
  SET_CHECKERS_IN_PROJECT_INFORMATION_REDUCER: "SET_CHECKERS_IN_PROJECT_INFORMATION_REDUCER",
  CLEAR_PROJECT_INFORMATION_REDUCER: "CLEAR_PROJECT_INFORMATION_REDUCER",
  // online mode reducer / actions
  RESET_ONLINE_MODE_WARNING_ALERT: "RESET_ONLINE_MODE_WARNING_ALERT",
  // WORD ALIGNMENT DATA
  UPDATE_ALIGNMENT_DATA: "UPDATE_ALIGNMENT_DATA",
  CLEAR_ALIGNMENT_DATA: "CLEAR_ALIGNMENT_DATA",
  //merge conflict
  CLEAR_MERGE_CONFLICTS_REDUCER: "CLEAR_MERGE_CONFLICTS_REDUCER",
  MERGE_CONFLICTS_CHECK: "MERGE_CONFLICTS_CHECK",
  //locale actions
  LOCALE_LOADED: "LOCALE_LOADED",
  // settings actions
  SET_CSV_SAVE_LOCATION: "SET_CSV_SAVE_LOCATION",
  SET_USFM_SAVE_LOCATION: "SET_USFM_SAVE_LOCATION",
  UPDATE_TOOL_SETTINGS: "UPDATE_TOOL_SETTINGS",
  SET_SETTING: "SET_SETTING",
  TOGGLE_SETTING: "TOGGLE_SETTING",
  // invalidated reducer
  SET_INVALIDATED: "SET_INVALIDATED",
  SET_INVALIDATED_CHECKS_TOTAL: "SET_INVALIDATED_CHECKS_TOTAL",
  SET_VERSE_EDITS_TOTAL: "SET_VERSE_EDITS_TOTAL",
  SET_INVALIDATED_ALIGNMENTS_TOTAL: 'SET_INVALIDATED_ALIGNMENTS_TOTAL',
  ADD_TOOL_API: 'ADD_TOOL_API'
};

export default ActionTypes;
