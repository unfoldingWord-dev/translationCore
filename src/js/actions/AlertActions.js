import types from './ActionTypes';

export const closeAlert = (id) => ({
  type: types.CLOSE_ALERT,
  id,
});

/**
 * Opens a new alert.
 * @param {string} id - the alert id
 * @param {*} message - the message to display in the alert
 * @param {func} onConfirm - custom confirm handler
 * @param {func} onCancel - custom cancel handler
 * @param {string} confirmText - the confirm button text
 * @param {string} cancelText - the cancel button text
 * @return {*}
 */
export const openAlert = (id, message, {
  onConfirm = null,
  onCancel = null,
  confirmText = null,
  cancelText = null,
} = {}) => ({
  type: types.OPEN_ALERT,
  id,
  message,
  onConfirm,
  confirmText,
  onCancel,
  cancelText,
});

/**
 * Opens a new alert that can be ignored.
 * @param {string} id - the alert id
 * @param {*} message
 * @param {func} onConfirm - custom confirm handler
 * @param {func} onCancel - custom cancel handler
 * @param {func} onIgnore - custom ignore handler
 * @param {string} confirmText - the confirm button text
 * @param {string} cancelText - the cancel button text
 * @param {string} ignoreText - the ignore checkbox text
 * @return {*}
 */
export const openIgnorableAlert = (id, message, {
  onConfirm = null,
  onCancel = null,
  onIgnore = null,
  confirmText = null,
  cancelText = null,
  ignoreText = null,
} = {}) => ({
  type: types.OPEN_ALERT,
  id,
  message,
  onConfirm,
  confirmText,
  onCancel,
  cancelText,
  onIgnore: typeof onIgnore === 'function' ? onIgnore : () => {},
  ignoreText,
});

/**
 * Ignores a specific alert for the rest of the session
 * @param {string} id - the alert id to ignore
 * @param {boolean} ignore - ignores or enables the alert
 * @return {*}
 */
export const ignoreAlert = (id, ignore = true) => ({
  type: types.IGNORE_ALERT,
  id,
  ignore,
});
