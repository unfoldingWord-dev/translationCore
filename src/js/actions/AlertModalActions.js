/**
 * @module Actions/AlertModal
 */

import consts from './ActionTypes';

/**
 * @description opens the alert dialog with the specified alert message.
 * @param {String} alertMessage - message to be displayed inside the alert dialog.
 * @param {Boolean} loading - true displays spinning icon and no action button in dialog.
 * @return {object} action content.
 */
export function openAlertDialog(alertMessage, loading) {
  return {
    type: consts.OPEN_ALERT_DIALOG,
    alertMessage,
    loading,
  };
}

/**
 * @description opens the option dialog with the specified alert message, and options.
 * @param {String} alertMessage - message to be displayed inside the alert dialog.
 * @param {function} callback - callback function that is called when clicking on button.  First parameter is text on button. Exception is when callback2 is given.
 * @param {String} button1Text - button text to show right button.
 * @param {String} button2Text - button text to show left of right button. (optional - if not present button is not added)
 * @param {String} buttonLinkText - button text to show on left link button. (optional - if not present button is not added)
 * @param {function} callback2 - optional callback function that is called user clicks on button2Text.
 * @return {Object} action content.
 */
export function openOptionDialog(alertMessage, callback, button1Text, button2Text, buttonLinkText = null, callback2 = null) {
  return {
    type: consts.OPEN_OPTION_DIALOG,
    alertMessage,
    callback,
    callback2,
    button1Text,
    button2Text,
    buttonLinkText,
  };
}

/**
 * @description closes the alert dialog.
 * @return {object} action content.
 */
export function closeAlertDialog() {
  return { type: consts.CLOSE_ALERT_DIALOG };
}
