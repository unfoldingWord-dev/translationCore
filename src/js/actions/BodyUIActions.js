import {
  getHomeScreenStep,
  getActiveHomeScreenSteps
} from '../selectors';
import types from './ActionTypes';

/**
 * @description toggles the home view based on param.
 * @param {boolean} boolean - true or false either shows or hides it.
 */
export const toggleHomeView = boolean => ({
  type: types.TOGGLE_HOME_VIEW,
  boolean
});

export const toggleWelcomeSplash = () => ({
  type: types.TOGGLE_WELCOME_SPLASH
});

export const goToNextStep = () => {
  return ((dispatch, getState) => {
    const stepIndex = getHomeScreenStep(getState());
    dispatch(goToStep(stepIndex + 1));
  });
};

export const goToPrevStep = () => {
  return ((dispatch, getState) => {
    const stepIndex = getHomeScreenStep(getState());
    dispatch(goToStep(stepIndex - 1));
  });
};

/**
 * Goes to specified step
 * @param {number} stepNumber - Number of step to go to in home stepper
 */
export const goToStep = stepNumber => {
  return ((dispatch, getState) => {
    if (stepNumber >= 0 && stepNumber <= 3) {
      const activeSteps = getActiveHomeScreenSteps(getState());
      if (!activeSteps[stepNumber]) return;
      dispatch({
        type: types.GO_TO_STEP,
        stepIndex: stepNumber
      });
    } else if (stepNumber < 0) {
      console.error('The min number of steps is 0. (0-3)');
    } else {
      console.error('The max number of steps is 3. (0-3)');
    }
  });
};

export const toggleProjectsFAB = () => ({
  type: types.TOGGLE_PROJECTS_FAB
});

export const openOnlineImportModal = () => ({
  type: types.OPEN_ONLINE_IMPORT_MODAL
});

export const closeOnlineImportModal = () => {
  return {
    type: types.CLOSE_ONLINE_IMPORT_MODAL
  };
};

export const openLicenseModal = () => ({
  type: types.OPEN_LICENSE_MODAL
});

export const closeLicenseModal = () => ({
  type: types.CLOSE_LICENSE_MODAL
});

/**
 * @description show or not show dimmed screen.
 * @param {bool} enable dims the screen if enabled otherwise removes the dim
 */
export const dimScreen = (enable=true) => ({
  type: types.SHOW_DIMMED_SCREEN,
  bool: enable
});
