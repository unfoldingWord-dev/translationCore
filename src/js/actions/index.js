/**
 * @author Ian Hoegen
 * @description: This serves as the new store and events api.
 * @usage:
 * Actions.sendAction('YOUR_EVENT', data) // Emits an event of type YOUR_EVENT, adds data to store
 * Actions.getData('YOUR_EVENT') // Retrieves data from the event emitted
 ***********************************************************************************/

const EventEmitter = require('events').EventEmitter;
const Reducers = require('../reducers');

class Actions extends EventEmitter {
  constructor() {
    super();
    this.store = [];
  }
  sendAction(eventName, actionObj) {
    if(Reducers[eventName]) {
      this.store[eventName] = Reducers[eventName](this.store[eventName], actionObj);
    } else {
      this.store[eventName] = actionObj;
    }
    this.emitChange(eventName);
  }

  subscribe(eventName, callback) {
    this.on(eventName, callback);
  }

  unsubscribe(eventName, callback) {
    this.removeListener(eventName, callback);
  }

  emitChange(eventName, data) {
    this.emit(eventName, data);
  }

  getData (actionType) {
    return this.store[actionType];
  }
}

var actions = new Actions();
module.exports = actions;
