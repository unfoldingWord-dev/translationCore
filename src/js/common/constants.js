import path from "path-extra";
import ospath from "ospath";
import packagefile from '../../../package.json';

export const APP_VERSION = packagefile.version;
export const MIN_COMPATIBLE_VERSION = packagefile.minCompatibleVersion;

export const TRANSLATION_WORDS = 'translationWords';
export const TRANSLATION_NOTES = 'translationNotes';
export const TRANSLATION_ACADEMY = 'translationAcademy';
export const TRANSLATION_HELPS = 'translationHelps';
export const WORD_ALIGNMENT = 'wordAlignment';

export const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');
export const USER_RESOURCES_PATH = path.join(ospath.home(), "translationCore", "resources");
export const THELPS_EN_RESOURCES_PATH = path.join(USER_RESOURCES_PATH, 'en', TRANSLATION_HELPS);
export const IMPORTS_PATH = path.join(ospath.home(), 'translationCore', 'imports');
export const PROJECT_INDEX_FOLDER_PATH = path.join('.apps', 'translationCore', 'index');
export const TEMP_IMPORT_PATH = path.join(ospath.home(), 'translationCore', 'imports', 'temp');

export const PROJECT_DOT_APPS_PATH = path.join('.apps', 'translationCore');
export const PROJECT_CHECKDATA_DIRECTORY = path.join(PROJECT_DOT_APPS_PATH, 'checkData');
export const STATIC_RESOURCES_PATH = path.join(__dirname, "../../../tcResources");

export const TC_VERSION = "tc_version";
export const SOURCE_CONTENT_UPDATER_MANIFEST = "source-content-updater-manifest.json";

export const tc_EDIT_VERSION_KEY = "tc_edit_version"; // do not change this string as it will break compatibility with other versions
export const tc_MIN_COMPATIBLE_VERSION_KEY = "tc_min_compatible_version"; // do not change this string as it will break compatibility with other versions
export const tc_MIN_VERSION_ERROR = "TC_MIN_VERSION_ERROR";

export const ORIGINAL_LANGUAGE = "originalLanguage";
export const TARGET_LANGUAGE = "targetLanguage";
export const TARGET_BIBLE = "targetBible";

export const toolCardCategories = {
  'kt': 'Key Terms',
  'names': 'Names',
  'other_terms': 'Other Terms',

  'culture': 'Culture',
  'figures': 'Figures of Speech',
  'numbers': 'Numbers',
  'discourse': 'Discourse',
  'grammar': 'Grammar',
  'other': 'Other'
};
