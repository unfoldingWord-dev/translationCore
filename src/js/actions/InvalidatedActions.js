import consts from './ActionTypes';
import isEqual from 'deep-equal';
import path from 'path-extra';
import fs from 'fs-extra';
import { checkSelectionOccurrences } from 'selections';
// Actions
import * as TargetLanguageActions from './TargetLanguageActions';
// helpers
import {generateTimestamp} from '../helpers/index';
import * as gatewayLanguageHelpers from '../helpers/gatewayLanguageHelpers';
import * as generatePathsUtil from '../common/generatePathsUtil';
import * as invalidatedCheckHelpers from '../helpers/invalidatedCheckHelpers';
// selectors
import {getUsername} from '../selectors';

/**
 * @description sets invalidatedReducer to true or false
 * @param {String} userName - The username of who invalidated.
 * @param {String} timestamp
 * @param {Boolean} invalidated - new state for invalidated flag
 * @return {object} action state.
 */
export function set(userName, timestamp, invalidated) {
  return ((dispatch, getState) => {
    const {
      gatewayLanguageCode,
      gatewayLanguageQuote
    } = gatewayLanguageHelpers.getGatewayLanguageCodeAndQuote(getState());

    dispatch({
      type: consts.SET_INVALIDATED,
      modifiedTimestamp: timestamp,
      gatewayLanguageCode,
      gatewayLanguageQuote,
      userName,
      invalidated
    });
  });
}

/**
 * @description sets invalidated to true or false in both invalidatedReducer and groupDataReducer
 * @param {String} username - The username of who invalidated.
 * @param {Boolean} invalidated - new state for invalidated flag
 * @return {object} action state.
 */
export const setInvalidated = (username, invalidated) => {
  return ((dispatch, getState) => {
    let state = getState();
    let contextId = state.contextIdReducer.contextId;

    dispatch(set(username, generateTimestamp(), invalidated));
    dispatch({
      type: consts.SET_INVALIDATION_IN_GROUPDATA,
      contextId,
      boolean: invalidated
    });
  });
};

export const getAllInvalidatedChecksForCurrentProject = () => {
  return ((dispatch, getState) => {
    const { projectSaveLocation, manifest: { project } } = getState().projectDetailsReducer;
    const bookAbbreviation = project.id;
    const invalidatedFolderPath = generatePathsUtil.getCheckDataFolderPath(projectSaveLocation, bookAbbreviation, 'invalidated');
    const verseEditFolderPath = generatePathsUtil.getCheckDataFolderPath(projectSaveLocation, bookAbbreviation, 'verseEdits');
    const verseEditsTotal = invalidatedCheckHelpers.getTotalOfEditedVerses(verseEditFolderPath);
    const invalidatedChecksTotal = invalidatedCheckHelpers.loadTotalOfInvalidatedChecksForCurrentProject(invalidatedFolderPath);
    const invalidatedAlignmentsTotal = invalidatedCheckHelpers.getTotalInvalidatedAlignments(projectSaveLocation, bookAbbreviation);

    dispatch({
      type: consts.SET_INVALIDATED_CHECKS_TOTAL,
      invalidatedChecksTotal
    });

    dispatch({
      type: consts.SET_VERSE_EDITS_TOTAL,
      verseEditsTotal
    });

    dispatch({
      type: consts.SET_INVALIDATED_ALIGNMENTS_TOTAL,
      invalidatedAlignmentsTotal
    });
  });
};

export const findInvalidatedSelectionsForAllCheckData = () => {
  return ((dispatch, getState) => {
    let state = getState();
    const projectPath = state.projectDetailsReducer.projectSaveLocation;
    const bookId = state.projectDetailsReducer.manifest.project.id;
    const selectionsPath = path.join(projectPath, '.apps', 'translationCore', 'checkData', 'selections', bookId);
    if (fs.existsSync(selectionsPath)) {
      let chapters = fs.readdirSync(selectionsPath);  
      for (let chapIdx in chapters) {
        let chapter = parseInt(chapters[chapIdx]);
        let chapterPath = path.join(selectionsPath, chapter.toString());
        dispatch(TargetLanguageActions.loadTargetLanguageChapter(chapter));
        state = getState();
        if (state.resourcesReducer && state.resourcesReducer.bibles && state.resourcesReducer.bibles.targetLanguage && state.resourcesReducer.bibles.targetLanguage.targetBible) {
          const bibleChapter = state.resourcesReducer.bibles.targetLanguage.targetBible[chapter];
          if (bibleChapter) {
            let verses = fs.readdirSync(chapterPath);
            for (let verseIdx in verses) {
              let verse = parseInt(chapters[verseIdx]);
              let versePath = path.join(chapterPath, verse.toString());
              const verseText = bibleChapter[verse];
              let files = fs.readdirSync(versePath);
              files = files.filter(file => { // filter the filenames to only use .json
                return path.extname(file) === '.json';
              });
              const sorted = files.sort().reverse(); // sort the files to use latest
              const done = {};
              for(let sortedIdx in sorted) {
                const filename = sorted[sortedIdx];
                const selectionsData = fs.readJsonSync(path.join(versePath, filename));
                const doneKey = selectionsData.contextId.tool + '-' + selectionsData.contextId.groupId + '-' + selectionsData.contextId.quote + '-' + selectionsData.contextId.occurrence;
                if ( ! done[doneKey]) {
                  const validSelections = checkSelectionOccurrences(verseText, selectionsData.selections);
                  if (!isEqual(selectionsData.selections, validSelections)) { // if true found invalidated check
                    const username = getUsername(state);
                    const modifiedTimestamp = generateTimestamp();
                    const invalidted = {
                      contextId: selectionsData.contextId,
                      invalidated: true,
                      userName: username,
                      modifiedTimestamp: modifiedTimestamp,
                      gatewayLanguageCode: selectionsData.gatewayLanguageCode,
                      gatewayLanguageQuote: selectionsData.gatewayLanguageQuote
                    };
                    const newFilename = modifiedTimestamp + '.json';
                    const invalidatedCheckPath = path.join(projectPath, '.apps', 'translationCore', 'checkData', 'invalidated', bookId, chapter.toString(), verse.toString());
                    fs.outputJSONSync(path.join(invalidatedCheckPath, newFilename.replace(/[:"]/g, '_')), invalidted);
                    done[doneKey] = true;
                  }
                }
              }
            }
          }
        }
      }
    }
  });
};
