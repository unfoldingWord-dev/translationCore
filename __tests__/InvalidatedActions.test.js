import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as InvalidatedActions from '../src/js/actions/InvalidatedActions';
import { TRANSLATION_WORDS } from '../src/js/common/constants';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('../src/js/helpers/gatewayLanguageHelpers', () => ({
  getGatewayLanguageCodeAndQuote: () => ({
    gatewayLanguageCode: 'en',
    gatewayLanguageQuote: 'authority',
  }),
}));

describe('InvalidatedActions.set', () => {
  test('set Invalidated true', () => {
    const invalidated = true;
    const store = mockStore({
      projectDetailsReducer: { manifest: { toolsSelectedGLs: { translationWords: 'en' } } },
      toolsReducer: { selectedTool: TRANSLATION_WORDS },
      groupsIndexReducer: {
        groupsIndex: [
          {
            id: 'apostle',
            name: 'apostle, apostles, apostleship',
          },
          {
            id: 'authority',
            name: 'authority, authorities',
          },
        ],
      },
      contextIdReducer: { contextId: { groupId: 'authority' } },
    });
    store.dispatch(InvalidatedActions.set('mannycolon', '2017-10-27T18:13:41.455Z', invalidated));
    const actions = store.getActions();
    expect(actions).toMatchSnapshot();
  });

  test('set Invalidated false', () => {
    const invalidated = false;
    const store = mockStore({
      projectDetailsReducer: { manifest: { toolsSelectedGLs: { translationWords: 'en' } } },
      toolsReducer: { selectedTool: TRANSLATION_WORDS },
      groupsIndexReducer: {
        groupsIndex: [
          {
            id: 'apostle',
            name: 'apostle, apostles, apostleship',
          },
          {
            id: 'authority',
            name: 'authority, authorities',
          },
        ],
      },
      contextIdReducer: { contextId: { groupId: 'authority' } },
    });
    store.dispatch(InvalidatedActions.set('mannycolon', '2017-10-27T18:13:41.455Z', invalidated));
    const actions = store.getActions();
    expect(actions).toMatchSnapshot();
  });
});

describe('InvalidatedActions.setInvalidated', () => {
  test('set Invalidated true', () => {
    const invalidated = true;
    const store = mockStore({
      projectDetailsReducer: { manifest: { toolsSelectedGLs: { translationWords: 'en' } } },
      toolsReducer: { selectedTool: TRANSLATION_WORDS },
      groupsIndexReducer: {
        groupsIndex: [
          {
            id: 'apostle',
            name: 'apostle, apostles, apostleship',
          },
          {
            id: 'authority',
            name: 'authority, authorities',
          },
        ],
      },
      contextIdReducer: { contextId: { groupId: 'authority' } },
    });
    store.dispatch(InvalidatedActions.setInvalidated('mannycolon', invalidated));
    const actions = store.getActions();
    expect(cleanOutDates(actions)).toMatchSnapshot();
  });

  test('set Invalidated false', () => {
    const invalidated = false;
    const store = mockStore({
      projectDetailsReducer: { manifest: { toolsSelectedGLs: { translationWords: 'en' } } },
      toolsReducer: { selectedTool: TRANSLATION_WORDS },
      groupsIndexReducer: {
        groupsIndex: [
          {
            id: 'apostle',
            name: 'apostle, apostles, apostleship',
          },
          {
            id: 'authority',
            name: 'authority, authorities',
          },
        ],
      },
      contextIdReducer: { contextId: { groupId: 'authority' } },
    });
    store.dispatch(InvalidatedActions.setInvalidated('mannycolon', invalidated));
    const actions = store.getActions();
    expect(cleanOutDates(actions)).toMatchSnapshot();
  });
});

//
// helpers
//

function cleanOutDates(actions) {
  const cleanedActions = JSON.parse(JSON.stringify(actions));

  for (let action of cleanedActions) {
    if (action.modifiedTimestamp) {
      delete action.modifiedTimestamp;
    }
  }
  return cleanedActions;
}

