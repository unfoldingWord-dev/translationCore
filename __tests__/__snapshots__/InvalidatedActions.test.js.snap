// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvalidatedActions.set set Invalidated false 1`] = `
Array [
  Object {
    "gatewayLanguageCode": "en",
    "gatewayLanguageQuote": "authority, authorities",
    "invalidated": false,
    "modifiedTimestamp": "2017-10-27T18:13:41.455Z",
    "type": "SET_INVALIDATED",
    "userName": "mannycolon",
  },
]
`;

exports[`InvalidatedActions.set set Invalidated true 1`] = `
Array [
  Object {
    "gatewayLanguageCode": "en",
    "gatewayLanguageQuote": "authority, authorities",
    "invalidated": true,
    "modifiedTimestamp": "2017-10-27T18:13:41.455Z",
    "type": "SET_INVALIDATED",
    "userName": "mannycolon",
  },
]
`;

exports[`InvalidatedActions.setInvalidated set Invalidated false 1`] = `
Array [
  Object {
    "gatewayLanguageCode": "en",
    "gatewayLanguageQuote": "authority, authorities",
    "invalidated": false,
    "type": "SET_INVALIDATED",
    "userName": "mannycolon",
  },
  Object {
    "boolean": false,
    "contextId": Object {
      "groupId": "authority",
    },
    "type": "SET_INVALIDATION_IN_GROUPDATA",
  },
]
`;

exports[`InvalidatedActions.setInvalidated set Invalidated true 1`] = `
Array [
  Object {
    "gatewayLanguageCode": "en",
    "gatewayLanguageQuote": "authority, authorities",
    "invalidated": true,
    "type": "SET_INVALIDATED",
    "userName": "mannycolon",
  },
  Object {
    "boolean": true,
    "contextId": Object {
      "groupId": "authority",
    },
    "type": "SET_INVALIDATION_IN_GROUPDATA",
  },
]
`;
